### PR TITLE
[torch.compile] Add compile-only mode

### PR DESCRIPTION
**Pull request by zou3519**
Summary
=======

This PR is on the way to overlapping torch.compile and weight loading. See [design doc]([https://docs.google.com/document/d/1hssZeQv_lJlKqOr0vfpoqEUNn4ASGHVRB9a49yhcY6E/edit?tab=t.0](https://docs.google.com/document/d/1hssZeQv_lJlKqOr0vfpoqEUNn4ASGHVRB9a49yhcY6E/edit?tab=t.0)) and [proof-of-concept PR](https://github.com/vllm-project/vllm/pull/36072) and [RFC](https://github.com/vllm-project/vllm/issues/34956)

- Adds `vllm compile <model> [options]` CLI command and `vllm.compile_model()` Python API
- These APIs populate vLLM's torch.compile cache and do nothing else. A subsequent `vllm serve` call can read from the cache and perform a warm start.
- They also use minimal GPU memory. This is accomplished through a combination of using FakeTensors (tensors with no storage that report device correctly) and Meta tensors (tensors with no storage that report device="meta"). Note that there is still some minimal GPU memory allocation (< 10 Mb, from GPUModelRunner runtime buffers), I did not go track down all of it, but I'm also not sure it matters.
- In the future we can extend "vllm compile" to more than just torch.compile; for example, if vLLM uses triton kernels, or JIT'ed flashinfer kernels, `vllm compile` may also just compile those and saved the compiled artifacts somewhere.

How it works
============
- `FakeModelLoader`. It initializes weights on `meta` device and runs weight post-processing on meta tensors. In the future this will probably need to wrap a real model loader but we don't need that for the current test cases.
- Before torch.compile tracing, `swap_meta_params_to_fake()` converts meta params to FakeTensors. This is required so that torch.compile sees Tensors with the correct devices.
- In theory we should also get the FakeModelLoader to give us FakeTensors, but FakeTensors do not yet support the type of Tensor subclasses that vLLM uses.
- We raise the `CompilationDone` exception after cache artifacts are saved to avoid executing with fake tensors. (calling torch.compile performs both the compilation and an initial run of invoking the compiled artifact with the inputs)
- `EngineCore` early-returns after `compile_or_warm_up_model`, skipping KV cache allocation, scheduler, and sampler setup

Test plan
=========
- Added tests for compile_only cold start followed by a warm start. The tests verify that the compile_only cold start uses no GPU memory, and the warm start does end up reading from the cache.

Future work
===========
In the following order:
- add an option to overlap torch.compile and weight loading. The main process will do weight loading while spawning a new process to do compile-only work (that does not use gpu memory)
- Extend this design to more weight loading schemes. For example, we currently support no weight processing. This will involve getting the additional weight processing to support meta tensors.

### CODE REVIEW

File: vllm/compilation/decorators.py [hunk 1]
```diff
@@ -599,9 +599,23 @@ def patched_inline_call(self_: Any) -> Any:
...
                     self.save_aot_compiled_function()
 
+                # In compile-only mode, raise CompilationDone after all
+                # piecewise graphs are compiled and cache artifacts saved.
+                # This is caught in gpu_worker.compile_or_warm_up_model()
+                # to skip execution with fake tensors.
+                if self.compilation_config.compile_only:
+                    from .backends import CompilationDone
+
+                    raise CompilationDone
```
[the remark this answers is not in the document]
Comment (zou3519) [reply vllm/compilation/decorators.py:609]: multimodal models are out of scope right now

File: vllm/compilation/decorators.py [hunk 2]
```diff
@@ -599,9 +599,23 @@ def patched_inline_call(self_: Any) -> Any:
...
+
                 with monitor_profiling_run():
                     output = self.aot_compiled_fn(self, *args, **kwargs)
             else:
+                # Same as above for non-AOT path.
+                if self.compilation_config.compile_only:
+                    from .backends import CompilationDone
+
+                    raise CompilationDone
```
[the remark this answers is not in the document]
Comment (zou3519) [reply vllm/compilation/decorators.py:618]: multimodal models are out of scope right now

File: vllm/model_executor/model_loader/fake_loader.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                dtype=param.dtype,
+                device=device,
+            )
+            # Navigate to the owning module and replace the parameter.
+            *path, attr = name.split(".")
+            parent = model.get_submodule(".".join(path)) if path else model
+            parent.register_parameter(
+                attr,
+                nn.Parameter(fake_data, requires_grad=param.requires_grad),
+            )
```
[the remark this answers is not in the document]
Comment (zou3519) [reply vllm/model_executor/model_loader/fake_loader.py:119]: this is intentional and the model behavior does not change based on the parameter type.

File: vllm/v1/worker/gpu_worker.py [hunk 4]
```diff
@@ -573,10 +578,59 @@ def compile_or_warm_up_model(self) -> float:
...
+
+            # In the normal path, the first torch.compile is triggered
+            # by profile_run() which calls _dummy_run(max_num_tokens).
+            # In compile-only mode we skip _initialize_kv_caches (which
+            # calls profile_run), so call it here to trigger compilation.
+            # CompilationDone is raised after vLLM's torch.compile cache
+            # and AOT artifact are saved, to prevent execution with fake
+            # tensors.
+            with suppress(CompilationDone):
+                self.model_runner.profile_run()
```
[the remark this answers is not in the document]
Comment (zou3519) [reply vllm/v1/worker/gpu_worker.py:611]: again, this is intentional
